File: backend/ml/predict.js
// Runs the custom-trained scam classifier (ml/src/model.py, exported to ONNX)
// against a job posting. Used as an additional signal fed into the Claude
// prompt in /api/analyze-job — never as a standalone verdict, since it was
// found to false-positive on short/sparse postings the extension's scraper
// typically produces (missing company_profile/requirements/benefits/logo
// fields it was trained with).
const fs = require('fs');
const path = require('path');
const ort = require('onnxruntime-node');
const { encodeText } = require('./tokenizer');

const ARTIFACTS_DIR = path.join(__dirname, 'artifacts');

let sessionPromise = null;
let vocab, catMaps, config;

function loadArtifacts() {
  vocab = JSON.parse(fs.readFileSync(path.join(ARTIFACTS_DIR, 'vocab.json'), 'utf8'));
  catMaps = JSON.parse(fs.readFileSync(path.join(ARTIFACTS_DIR, 'cat_maps.json'), 'utf8'));
  config = JSON.parse(fs.readFileSync(path.join(ARTIFACTS_DIR, 'config.json'), 'utf8'));
}

function getSession() {
  if (!sessionPromise) {
    loadArtifacts();
    sessionPromise = ort.InferenceSession.create(path.join(ARTIFACTS_DIR, 'model.onnx'));
  }
  return sessionPromise;
}

function sigmoid(x) {
  return 1 / (1 + Math.exp(-x));
}

// Maps the extension's scraped jobData shape onto the schema the model was
// trained on. Fields the extension doesn't scrape (company_profile,
// requirements, benefits, employment_type, industry, has_company_logo, ...)
// are left at their "missing" training-time value.
function buildFeatures(jobData) {
  const fullText = [jobData.title || '', jobData.description || ''].join(' ');

  const catCols = config.cat_cols;
  const catValues = catCols.map(() => '__missing__');

  const binFeatures = [
    0, // telecommuting - not scraped
    0, // has_company_logo - not scraped
    0, // has_questions - not scraped
    jobData.salary ? 1 : 0, // has_salary
  ];

  return { fullText, catCols, catValues, binFeatures };
}

async function scoreJobPosting(jobData) {
  const session = await getSession();
  const { fullText, catCols, catValues, binFeatures } = buildFeatures(jobData);

  const { ids, attn } = encodeText(fullText, vocab, config.max_len);
  const catIds = catCols.map((col, i) => {
    const map = catMaps[col];
    const val = catValues[i];
    return BigInt(val in map ? map[val] : 0);
  });

  const feeds = {
    input_ids: new ort.Tensor('int64', BigInt64Array.from(ids.map(BigInt)), [1, ids.length]),
    attn_mask: new ort.Tensor('int64', BigInt64Array.from(attn.map(BigInt)), [1, attn.length]),
    cat_ids: new ort.Tensor('int64', BigInt64Array.from(catIds), [1, catIds.length]),
    bin_features: new ort.Tensor('float32', Float32Array.from(binFeatures), [1, binFeatures.length]),
  };

  const results = await session.run(feeds);
  const logit = results.logit.data[0];
  const probability = sigmoid(logit);

  return { probability };
}

module.exports = { scoreJobPosting };

File: ml/src/predict.py
"""
Score a single job posting with the trained model.

Usage:
    python3 predict.py --title "..." --description "..." [--company_profile ...] [--requirements ...] \\
        [--benefits ...] [--employment_type "Full-time"] [--required_experience "Mid-Senior level"] \\
        [--required_education "Bachelor's Degree"] [--industry "..."] [--function "..."] \\
        [--telecommuting] [--has_company_logo] [--has_questions] [--has_salary]
"""
import argparse
import json
import os

import torch

from model import ScamClassifier
from preprocess import encode_text, CAT_COLS
from train import ARTIFACTS_DIR, DEVICE


def load_model():
    with open(os.path.join(ARTIFACTS_DIR, "vocab.json")) as f:
        vocab = json.load(f)
    with open(os.path.join(ARTIFACTS_DIR, "cat_maps.json")) as f:
        cat_maps = json.load(f)
    with open(os.path.join(ARTIFACTS_DIR, "config.json")) as f:
        config = json.load(f)

    cat_cardinalities = [len(cat_maps[col]) for col in config["cat_cols"]]
    model = ScamClassifier(
        vocab_size=len(vocab),
        max_len=config["max_len"],
        cat_cardinalities=cat_cardinalities,
        num_bin_features=4,
    ).to(DEVICE)
    model.load_state_dict(torch.load(os.path.join(ARTIFACTS_DIR, "best_model.pt"), map_location=DEVICE))
    model.eval()
    return model, vocab, cat_maps, config


def predict(model, vocab, cat_maps, config, posting):
    full_text = " ".join(
        posting.get(col, "") or "" for col in ["title", "company_profile", "description", "requirements", "benefits"]
    )
    ids, attn = encode_text(full_text, vocab, config["max_len"])
    input_ids = torch.tensor([ids]).to(DEVICE)
    attn_mask = torch.tensor([attn]).to(DEVICE)

    cat_ids = torch.tensor(
        [[cat_maps[col].get(posting.get(col, "__missing__"), 0) for col in CAT_COLS]]
    ).to(DEVICE)
    bin_features = torch.tensor(
        [[
            float(posting.get("telecommuting", False)),
            float(posting.get("has_company_logo", False)),
            float(posting.get("has_questions", False)),
            float(posting.get("has_salary", False)),
        ]]
    ).to(DEVICE)

    with torch.no_grad():
        logit = model(input_ids, attn_mask, cat_ids, bin_features)
        prob = torch.sigmoid(logit).item()
    return prob


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--title", default="")
    parser.add_argument("--company_profile", default="")
    parser.add_argument("--description", default="")
    parser.add_argument("--requirements", default="")
    parser.add_argument("--benefits", default="")
    parser.add_argument("--employment_type", default="__missing__")
    parser.add_argument("--required_experience", default="__missing__")
    parser.add_argument("--required_education", default="__missing__")
    parser.add_argument("--industry", default="__missing__")
    parser.add_argument("--function", default="__missing__")
    parser.add_argument("--telecommuting", action="store_true")
    parser.add_argument("--has_company_logo", action="store_true")
    parser.add_argument("--has_questions", action="store_true")
    parser.add_argument("--has_salary", action="store_true")
    args = parser.parse_args()

    model, vocab, cat_maps, config = load_model()
    prob = predict(model, vocab, cat_maps, config, vars(args))

    print(f"fraud probability: {prob:.4f}")
    print("verdict:", "LIKELY SCAM" if prob >= 0.5 else "likely legitimate")


if __name__ == "__main__":
    main()
